### PR TITLE
Add details about nonbacktracking

### LINKED ISSUE #34174: What limitations does NonBacktracking have?
The new RegexOptions.NonBacktracking value seems to solve the problem of catastrophic backtracking, I'm not sure about that. But it has to come with some limitations of the patterns it can match. If this protection was free, it would be the default. But it's not. So when should I *not* use this option? Or when does it make a difference at all? In fact the documentation doesn't really tell what the option does. And should I still set a timeout with this option?

---
#### Document Details

⚠ *Do not edit this section. It is required for learn.microsoft.com ➟ GitHub issue linking.*

* ID: bf9b6611-d08f-2fe9-216b-e0658d7f1553
* Version Independent ID: 3d3be892-b9b5-ca0b-1193-9fa49716decb
* Content: [Options for regular expression](https://learn.microsoft.com/en-us/dotnet/standard/base-types/regular-expression-options)
* Content Source: [docs/standard/base-types/regular-expression-options.md](https://github.com/dotnet/docs/blob/main/docs/standard/base-types/regular-expression-options.md)
* Product: **dotnet-fundamentals**
* GitHub Login: @adegeo
* Microsoft Alias: **adegeo**

### PR DESCRIPTION
**Pull request by gewarren**
Fixes #34174 

---

#### Internal previews

| 📄 File | 🔗 Preview link |
|:--|:--|
| [docs/core/whats-new/dotnet-7.md](https://github.com/dotnet/docs/blob/385ad93468cf20e2ef2a6454f38342dc4674514a/docs/core/whats-new/dotnet-7.md) | [What's new in .NET 7](https://review.learn.microsoft.com/en-us/dotnet/core/whats-new/dotnet-7?branch=pr-en-us-35051) |
| [docs/standard/base-types/regular-expression-options.md](https://github.com/dotnet/docs/blob/385ad93468cf20e2ef2a6454f38342dc4674514a/docs/standard/base-types/regular-expression-options.md) | [Regular expression options](https://review.learn.microsoft.com/en-us/dotnet/standard/base-types/regular-expression-options?branch=pr-en-us-35051) |

### CODE REVIEW

File: docs/standard/base-types/regular-expression-options.md [hunk 1]
```diff
@@ -385,6 +385,21 @@ The following example is identical to the previous example, except that the stat
 [!code-csharp[Conceptual.Regex.Language.Options#15](../../../samples/snippets/csharp/VS_Snippets_CLR/conceptual.regex.language.options/cs/culture1.cs#15)]
 [!code-vb[Conceptual.Regex.Language.Options#15](../../../samples/snippets/visualbasic/VS_Snippets_CLR/conceptual.regex.language.options/vb/culture1.vb#15)]
 
+## Nonbacktracking mode
+
+By default, .NET's regex engine uses *backtracking* to try to find pattern matches. A backtracking engine is one that tries to match one pattern, and if that fails, goes backs and tries to match an alternate pattern, and so on. A backtracking engine is very fast for typical cases, but slows down exponentially as then number of pattern alternations increases, which can lead to *catastrophic backtracking*. The <xref:System.Text.RegularExpressions.RegexOptions.NonBacktracking?displayProperty=nameWithType> option doesn't use backtracking and avoids that worst-case scenario. While the nonbacktracking search is fast in all cases, it's not *always* faster than the default backtracking engine, and that isn't its goal.
```
Comment (gewarren) [inline docs/standard/base-types/regular-expression-options.md:390]:
```suggestion
By default, .NET's regex engine uses *backtracking* to try to find pattern matches. A backtracking engine is one that tries to match one pattern, and if that fails, goes backs and tries to match an alternate pattern, and so on. A backtracking engine is very fast for typical cases, but slows down exponentially as the number of pattern alternations increases, which can lead to *catastrophic backtracking*. The <xref:System.Text.RegularExpressions.RegexOptions.NonBacktracking?displayProperty=nameWithType> option doesn't use backtracking and avoids that worst-case scenario. While the nonbacktracking search is fast in all cases, it's not *always* faster than the default backtracking engine, and that isn't its goal.
```

File: docs/standard/base-types/regular-expression-options.md [hunk 2]
```diff
@@ -385,6 +385,21 @@ The following example is identical to the previous example, except that the stat
 [!code-csharp[Conceptual.Regex.Language.Options#15](../../../samples/snippets/csharp/VS_Snippets_CLR/conceptual.regex.language.options/cs/culture1.cs#15)]
 [!code-vb[Conceptual.Regex.Language.Options#15](../../../samples/snippets/visualbasic/VS_Snippets_CLR/conceptual.regex.language.options/vb/culture1.vb#15)]
 
+## Nonbacktracking mode
+
+By default, .NET's regex engine uses *backtracking* to try to find pattern matches. A backtracking engine is one that tries to match one pattern, and if that fails, goes backs and tries to match an alternate pattern, and so on. A backtracking engine is very fast for typical cases, but slows down exponentially as the number of pattern alternations increases, which can lead to *catastrophic backtracking*. The <xref:System.Text.RegularExpressions.RegexOptions.NonBacktracking?displayProperty=nameWithType> option doesn't use backtracking and avoids that worst-case scenario. While the nonbacktracking search is fast in all cases, it's not *always* faster than the default backtracking engine, and that isn't its goal.
```
Comment (stephentoub) [inline docs/standard/base-types/regular-expression-options.md:390]:
```suggestion
By default, .NET's regex engine uses *backtracking* to try to find pattern matches. A backtracking engine is one that tries to match one pattern, and if that fails, goes backs and tries to match an alternate pattern, and so on. A backtracking engine is very fast for typical cases, but slows down as the number of pattern alternations increases, which can lead to *catastrophic backtracking*. The <xref:System.Text.RegularExpressions.RegexOptions.NonBacktracking?displayProperty=nameWithType> option doesn't use backtracking and avoids that worst-case scenario. The goal of the non-backtracking option is to provide consistently good behavior, regardless of the input being searched.
```

File: docs/standard/base-types/regular-expression-options.md [hunk 3]
```diff
@@ -385,6 +385,21 @@ The following example is identical to the previous example, except that the stat
...
+
+The <xref:System.Text.RegularExpressions.RegexOptions.NonBacktracking?displayProperty=nameWithType> option doesn't support everything the other built-in engines support. In particular, the option can't be used in conjunction with <xref:System.Text.RegularExpressions.RegexOptions.RightToLeft?displayProperty=nameWithType> or <xref:System.Text.RegularExpressions.RegexOptions.ECMAScript?displayProperty=nameWithType>. It also doesn't allow for the following constructs in the pattern:
+
+- Atomic groups
+- Backreferences
+- Balancing groups
+- Conditional
```
Comment (stephentoub) [inline docs/standard/base-types/regular-expression-options.md:397]:
```suggestion
- Conditionals
```

**Review by BillWagner — APPROVED** [review]
This LGTM @gewarren 

Let's :shipit:

**Comment by gewarren** [comment]
Thanks for reviewing @stephentoub @BillWagner

**Pull request merged**